### PR TITLE
First draft of Normalize plugin

### PR DESCRIPTION
**Pull request by bhushankawadkar**
First draft of Normalize plugin added.
1. Plugin code added as per requirement and design.
2. JUnit test not added as failing due to error Caused by: java.io.IOException: Incomplete document in MockSink. Detail description emailed. Once resolve this issue will add the test file.

### CODE REVIEW

**Comment by bhushankawadkar** [comment]
First draft of Normalize plugin added. (please consider both commit [b70152e](https://github.com/caskdata/hydrator-plugins/pull/327/commits/b70152e6afcb2708db7c003afa2a92a907ecbc31 and [5e7766c]%28https://github.com/caskdata/hydrator-plugins/pull/327/commits/5e7766c6c073d5a4130bcc4bcff234795e83c100%29)
1. Plugin code added as per requirement and design.
2. JUnit test not added as failing due to error Caused by: java.io.IOException: Incomplete document in MockSink. Detail description emailed. Once resolve this issue will add the test file.

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
+    super.configurePipeline(pipelineConfigurer);
+    //validate fields with input schema
+    Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
+    //validate mapping fields
+    String[] fieldMappingArray = config.fieldMapping.split(",");
+    for (String fieldMapping : fieldMappingArray) {
+      String[] mappings = fieldMapping.split(":");
+      Preconditions.checkArgument(inputSchema.getField(mappings[0]) != null, "Mapping field '" + mappings[0]
```
Comment (nitinmotgi) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java]: If mappings is null, wouldn't this line through a exception ?

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        + "' not present in input schema.");
+    }
+    //validate normalizing fields
+    String[] fieldNormalizingArray = config.fieldNormalizing.split(",");
+    for (String fieldNormalizing : fieldNormalizingArray) {
+      String[] fields = fieldNormalizing.split(":");
+      Preconditions.checkArgument(inputSchema.getField(fields[0]) != null, "Normalizing field '" + fields[0]
```
Comment (nitinmotgi) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java]: Same here.

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    if (outputSchema != null) {
+      return;
+    }
+    //create output schema
+    List<Schema.Field> outputFields = Lists.newArrayList();
+    mappingFieldMap = new HashMap<String, String>();
+    String[] fieldMappingArray = config.fieldMapping.split(",");
+    for (String fieldMapping : fieldMappingArray) {
+      String[] mappings = fieldMapping.split(":");
+      Preconditions.checkArgument(mappings.length == 2, "Output schema field is missing for mapping field '" +
```
Comment (nitinmotgi) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java]: ditto

**Comment by nitinmotgi** [comment]
Minor comments other wise LGTM 👍

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java (hunk 1 above)
Comment (bhushankawadkar) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java]: Tested on hydrator UI, widget does not allow null or empty key. It throws error, hence never be null.

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java (hunk 2 above)
Comment (bhushankawadkar) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java]: Tested on hydratory UI, widget does not allow null or empty value, hence cannot be null.

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java (hunk 3 above)
Comment (bhushankawadkar) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java]: same here

**Comment by bhushankawadkar** [comment]
JUnit test cases added.
Travis build is failing due to ETLTPFSTestRun, so could not reach upto NormalizeTest class.
However it is building successfully on local machine.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Hydrator Plugin Collection ......................... SUCCESS [  2.332 s]
[INFO] Hydrator Common .................................... SUCCESS [  4.186 s]
[INFO] Transform Plugins .................................. SUCCESS [01:58 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:05 min
[INFO] Finished at: 2016-07-21T18:22:03+05:30
[INFO] Final Memory: 80M/825M
[INFO] ------------------------------------------------------------------------

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java (hunk 1 above)
Comment (nitinmotgi) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java]: You don't need to use Hydrator to create JSON. You can also create a JSON by hand. So, handling that situation is also important.

Comment (bhushankawadkar) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java]: sure I will update the code wherever required

**Comment by bhushankawadkar** [comment]
Review comments updated
1. added validation for null value check of mapping and normalizing fields
2. added test cases for null / empty mapping and normalizing fields

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+      Preconditions.checkArgument(validValueField.equals(fields[2]), "Value mapping is invalid for " +
+        "normalize field '" + fields[0] + "'. It must be same for all normalize fields.");
+      normalizeFieldList.add(fields[0]);
+      normalizeFieldMap.put(fields[0] + NAME_KEY_SUFFIX, fields[1]);
+      normalizeFieldMap.put(fields[0] + VALUE_KEY_SUFFIX, fields[2]);
+    }
+    outputFields.add(Schema.Field.of(validTypeField, Schema.of(Schema.Type.STRING)));
+    outputFields.add(Schema.Field.of(validValueField, Schema.of(Schema.Type.STRING)));
+    outputSchema = Schema.recordOf("outputSchema", outputFields);
+  }
```
Comment (nitinmotgi) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java:153]: @bhushankawadkar we should have a way to generate the schema from FE. Meaning use the service and integration with frontend with button to generate the schema or we should get user to enter output schema.

Comment (bhushankawadkar) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java:153]: @nitinmotgi Code updated to get output schema entered by user and validating it against mapping and normalizing fields.

**Comment by bhushankawadkar** [comment]
Review comments updated
1. Code updated to get output schema entered by user and validating it against mapping and normalizing fields.
2. Test cases added to validate output schema.

**Comment by nitinmotgi** [comment]
LGTM :+1:

**Pull request merged**